src/coreclr/nativeaot/Runtime/MiscHelpers.cpp

```diff
@@ -375,35 +375,40 @@ FCIMPL1(uint8_t *, RhGetCodeTarget, uint8_t * pCodeOrg)
 #elif TARGET_RISCV64
     uint32_t * pCode = (uint32_t *)pCodeOrg;
     // is this "addi a0, a0, 8"?
-    if (pCode[0] == 0x000200b3)  // Encoding for `addi a0, a0, 8` in 32-bit RISC-V
+    if (pCode[0] == 0x00850513)  // Encoding for `addi a0, a0, 8` in 32-bit instruction format
     {
         // unboxing sequence
         unboxingStub = true;
         pCode++;
     }
     // is this an indirect jump?
     // lui t0, imm; jalr t0, t0, imm12
-    if ((pCode[0] & 0x000ff000) == 0x00020000 &&
-        (pCode[1] & 0x000ff000) == 0x000000a0 &&
-        (pCode[2] & 0x0000000f) == 0x00000000)
+    if ((pCode[0] & 0x7f) == 0x17 &&                 // auipc
+        (pCode[1] & 0x707f) == 0x3003 &&             // ld with funct3=011
+        (pCode[2] & 0x707f) == 0x0067)               // jr (jalr with x0 as rd and funct3=000)
     {
-        // normal import stub - dist to IAT cell is relative to (PC & ~0xfff)
-        // lui: imm = SignExtend(imm20:Zeros(12), 64);
-        int64_t distToIatCell = ((((int64_t)pCode[0] & ~0xfff) << 12) >> 32);
-        // jalr: offset = SignExtend(imm12, 64);
-        distToIatCell += (int64_t)(pCode[1] & 0xfff);
+        // Compute the distance to the IAT cell
+        int64_t distToIatCell = (((int64_t)pCode[0] & 0xfffff000) >> 12);  // Extract imm20 from auipc
```

Comment:
```suggestion
        int64_t distToIatCell = (((int32_t)pCode[0]) >> 12) << 12;  // Extract imm20 from auipc
```
auipc's immediate is scaled by 4096 so you don't need to shift it down, just clear the lower 12 bits and replicate the sign bit